iup/__init__.py

```diff
@@ -1132,26 +1133,51 @@ def assert_column_name_all(estimate, column_name):
 
 
 ###### evaluation metrics #####
-def get_mspe(data: IncidentUptakeData, pred: PointForecast) -> pl.DataFrame:
+def check_date_match(data: IncidentUptakeData, pred: PointForecast):
     """
-    Calculate MSPE between data and pred
+    Check the dates between data and pred.
+    Dates must be 1-on-1 equal and no duplicate.
     ----------------------
-    Input: data, pred in a matching and validated format
-    Return: pl.DataFrame with MSPE and the forecast start date and end date
+
+    Parameters
+    data:
+        The observed data used for modeling. Should be IncidentUptakeData
+    pred:
+        The forecast made by model. Should be PointForecast
+
+    Return
+    Error if conditions fail to meet.
 
     """
-    # Check the conditions for date match:
-    # 1. Mutual dates must exist between data and prediction.
-    assert any(
-        data["date"].is_in(pred["date"])
-    ), "No matched dates between data and prediction."
+    # sort data and pred by date #
+    data = data.sort("date")
+    pred = pred.sort("date")
 
-    # 2. There should not be any duplicated date in either data or prediction.
-    common_dates = data.filter(pl.col("date").is_in(pred["date"])).select("date")
+    # 1. Dates must be 1-on-1 equal
+    testing.assert_series_equal(data["date"], pred["date"])
 
+    # 2. There should not be any duplicated date in either data or prediction.
     assert (
-        len(common_dates) == common_dates.n_unique()
-    ), "Duplicated dates are found in data or prediction."
+        len(data["date"]) == data["date"].n_unique()
+    ), "Duplicated dates are found in data and prediction."
+
+
+def get_mspe(data: IncidentUptakeData, pred: PointForecast) -> pl.DataFrame:
+    """
+    Calculate MSPE between observed data and forecast.
+    ----------------------
+
+    Parameters
+    data:
+        The observed data used for modeling. Should be IncidentUptakeData
+    pred:
+        The forecast made by model. Should be PointForecast
+
+    Return
+    pl.DataFrame with MSPE and the forecast start date and end date
+
+    """
+    check_date_match(data, pred)
 
     joined = data.join(pred, on="date", how="inner", validate="1:1")
```

Comment:
Check with @swo but I think he and I were both suggesting that this calculation of mspe should also be its own function, perhaps called `calculate_mpse`. In other words, `get_mspe` just calls two functions, `check_date_match` and `calculate_mspe` in that order. The advantage is that you can test `calculate_mspe` completely independent of testing `check_date_match` (whereas here you must `check_date_match` in order to test that mspe is calculated correctly).